lib/resourcebuilder/apps.go

```diff
@@ -72,6 +76,35 @@ func (b *deploymentBuilder) Do(ctx context.Context) error {
 		}
 	}
 
+	// if we detect the CVO deployment we need to replace the KUBERNETES_SERVICE_HOST env var with the internal load
+	// balancer to be resilient to kube-apiserver rollouts that cause the localhost server to become non-responsive for
+	// multiple minutes.
+	if deployment.Namespace == "openshift-cluster-version" && deployment.Name == "cluster-version-operator" {
+		infrastructureConfig, err := b.infrastructureGetter.Infrastructures().Get("cluster", metav1.GetOptions{})
+		// not found just means that we don't have infrastructure configuration yet, so we should tolerate not found and avoid substitution
+		if err != nil && !errors.IsNotFound(err) {
+			return err
+		}
+		if !errors.IsNotFound(err) {
+			lbURL, err := url.Parse(infrastructureConfig.Status.APIServerInternalURL)
+			if err != nil {
+				return err
+			}
+			// if we have any error and have empty strings, substitution below will do nothing and leave the manifest specified value
+			// errors can happen when the port is not specified, in which case we have a host and we write that into the env vars
+			lbHost, lbPort, _ := net.SplitHostPort(lbURL.Host)
```

Comment:
I don't think a blanket-ignore works.

```go
package main

import (
	"fmt"
	"net"
	"net/url"
)

func main() {
	config := "https://example.com"
	lbURL, err := url.Parse(config)
	fmt.Printf("%v %s\n", lbURL, err)
	lbHost, lbPort, err := net.SplitHostPort(lbURL.Host)
	fmt.Printf("%q %q %s\n", lbHost, lbPort, err)
}
```

[gives][1]:

```
https://example.com %!s(<nil>)
"" "" address example.com: missing port in address
```

```
lbHost, lbPort, err := net.SplitHostPort(lbURL.Host)
if err != nil {
  lbHost = lbURL.Host
}
```

Would be nice if there was an error detector for `missing port in address`, but I don't see one.  I'd be ok with:

```go
if err != nil {
  if strings.Contains(err.Error(), "missing port in address") {
    lbHost = lbURL.Host
  } else {
    return err
  }
}
```

too, if we just want to treat the substring as a committed API.


[1]: https://play.golang.org/p/stQUuNkqp3m

Reply:
f0f00ff9b9b18 has the substring check.  Marking this thread resolved.